### ISSUE TITLE
Support for using `or []` to default `Optional[List[T]]` to an empty `List[T]`

### ISSUE THREAD

**Issue by sbdchd** [microsoft/pyright#1441]
**Is your feature request related to a problem? Please describe.**

Currently writing `bar = foo or []` where `foo` is type `Optional[List[T]]` 
results in `bar` being typed as `List[T] | List[Any]`:

```python
from typing import Optional, List

def get_foo() -> Optional[List[str]]:
    ...

foo = get_foo()
bar = foo or []
reveal_type(bar) # Type of "bar" is "List[str] | list[Any]"
```

**Describe the solution you'd like**

I'd like `bar` in the above example to be typed `List[T]` instead of 
`List[T] | List[Any]`

**Additional context**

In actual code I've encountered this in two scenarios, one with the `__init__` 
method for a class that takes a `Optional[List[T]]`, and another more common
case where `or []` is used as a short hand for defaulting to an empty `list`.

The setups end up looking something like:

```python
from typing import Optional, List


class Foo:
    def __init__(self, items: Optional[List[int]] = None) -> None:
        self.items = items or []


reveal_type(Foo().items) # Type of "Foo().items" is "List[int] | list[Any]"


class Request:
    def get_enabled_features(self) -> Optional[List[str]]:
        ...


def func(request: Request) -> None:
    features = request.get_enabled_features() or []
    reveal_type(features) # Type of "features" is "List[str] | list[Any]"
```

pyright 1.1.74

**Comment by erictraut** [comment]
This would be inconsistent with the way Pyright performs type inference for variables. The rules Pyright uses are:
1. If a variable has a declared type, use it as the "expected type" and apply bidirectional type inference for any expression assigned to it.
2. If a variable doesn't have a declared type, its inferred type is the union of the types of expressions assigned to it.

Because you have not declared a type for `bar` in the above example, you are falling into case 2.

The correct workaround is to declare a type for the variable.
```python
bar: List[str] = foo or []
```

**Comment by erictraut** [comment]
Thinking about this a bit more, since `foo or []` is the same expression, it wouldn't be inconsistent with the normal inference rules. I'll make the change as you suggested.

BTW, are you really running pyright 1.1.74? That's four months and 33 releases out of date!

**Comment by sbdchd** [comment]
Thanks for the quick update, I didn't realize the version I was running was that old, yikes! 
I've upgraded to the latest and greatest.

**Comment by erictraut** [comment]
This is included in Pyright 1.1.108, which I just published. It will also be in the next release of Pylance.
